## Preprocessing

In [1]:
!pip install tensorflow

  Created wheel for termcolor: filename=termcolor-1.1.0-py3-none-any.whl size=4829 sha256=cc0ae6a292a9fbb43e13b4779ed6f04dc6ddeb3ceeb72df0f90a807d620f3e6a
  Stored in directory: c:\users\miss_\appdata\local\pip\cache\wheels\a0\16\9c\5473df82468f958445479c59e784896fa24f4a5fc024b0f501
Successfully built termcolor


In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop('EIN', axis = 1)
application_df = application_df.drop('NAME' , axis = 1)
application_df.head(5)

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

In [3]:
# Determine the number of unique values in each column.
unique_columns = application_df.nunique()
unique_columns

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts for binning
app_unique = application_df['APPLICATION_TYPE'].value_counts()
app_unique

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T14        3
T25        3
T15        2
T29        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = ['T9', 'T13', 'T12', 'T25', 'T14', 'T29', 'T15', 'T17']

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      260
T2          16
Name: APPLICATION_TYPE, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts for binning
class_unique = application_df['CLASSIFICATION'].value_counts()
class_unique

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C2561        1
C3700        1
C8210        1
C2150        1
C6100        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
class_unique_nonzero = class_unique[class_unique > 1]
class_unique_small = class_unique[class_unique < 1883]
df_class = pd.DataFrame(class_unique_small)
classifications_to_replace = df_class.index
classifications_to_replace

Index(['C7000', 'C1700', 'C4000', 'C5000', 'C1270', 'C2700', 'C2800', 'C7100',
       'C1300', 'C1280', 'C1230', 'C1400', 'C2300', 'C7200', 'C1240', 'C8000',
       'C7120', 'C1500', 'C1800', 'C6000', 'C1250', 'C8200', 'C1278', 'C1238',
       'C1235', 'C1237', 'C7210', 'C4100', 'C1720', 'C2400', 'C1257', 'C1600',
       'C0', 'C2710', 'C1260', 'C1267', 'C3200', 'C1256', 'C1234', 'C1246',
       'C1283', 'C1900', 'C1248', 'C1370', 'C2500', 'C1580', 'C2600', 'C4120',
       'C4200', 'C1728', 'C1820', 'C1236', 'C2570', 'C2190', 'C1732', 'C4500',
       'C2380', 'C1245', 'C2170', 'C5200', 'C1570', 'C2561', 'C3700', 'C8210',
       'C2150', 'C6100'],
      dtype='object')

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`


# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
with_dummies = pd.get_dummies(application_df)
with_dummies.head()

STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  \
0                     1                     0                    0   
1                     0                     0                    0   
2                     0                     0                    0   
3                     0                     0                    0   
4                     0                     0                    0   

   APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  ...  \
0                    0                    0                    0  ...   
1                    1                    0                    0  ...   
2                    0                    0                    1  ...   
3                    1                    0                    0  ...   
4                    1                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 45 columns]

In [10]:
# Split our preprocessed data into our features and target arrays
y = with_dummies['IS_SUCCESSFUL']
x = with_dummies.drop('IS_SUCCESSFUL', axis = 1)

# Split the preprocessed data into a training and testing dataset
x_train, x_test, y_train, y_test = train_test_split(x, y, random_state = 1)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(x_train)

# Scale the data
x_train_scaled = X_scaler.transform(x_train)
x_test_scaled = X_scaler.transform(x_test)

In [12]:
#len(x_train[0])
x_train

STATUS  ASK_AMT  APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  \
10679       1     5000                       0                     0   
5052        1     5000                       0                     0   
9990        1     5000                       0                     0   
25173       1     5000                       0                     0   
11405       1     5000                       0                     1   
...       ...      ...                     ...                   ...   
7813        1     5000                       0                     0   
32511       1     5000                       0                     0   
5192        1     5000                       0                     0   
12172       1     5000                       0                     0   
33003       1     5000                       0                     0   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T3  \
10679                     0                    0                    1   
5052                      0                    0                    0   
9990                      0                    0                    1   
25173                     0                    0                    1   
11405                     0                    0                    0   
...                     ...                  ...                  ...   
7813                      0                    0                    1   
32511                     0                    0                    0   
5192                      0                    0                    1   
12172                     0                    0                    1   
33003                     0                    0                    0   

       APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
10679                    0                    0                    0  ...   
5052                     0                    0                    1  ...   
9990                     0                    0                    0  ...   
25173                    0                    0                    0  ...   
11405                    0                    0                    0  ...   
...                    ...                  ...                  ...  ...   
7813                     0                    0                    0  ...   
32511                    0                    1                    0  ...   
5192                     0                    0                    0  ...   
12172                    0                    0                    0  ...   
33003                    1                    0                    0  ...   

       INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
10679                  0                       0                         0   
5052                   0                       0                         0   
9990                   0                       0                         0   
25173                  0                       0                         0   
11405                  0                       0                         0   
...                  ...                     ...                       ...   
7813                   0                       0                         0   
32511                  0                       0                         0   
5192                   0                       0                         0   
12172                  0                       0                         0   
33003                  0                       0                         0   

       INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
10679                   0                 0                       0   
5052                    0                 0                       0   
9990                    0                 0                       0   
25173                   0                 0                       0   
11405                   0                 0     

## Compile, Train and Evaluate the Model

In [13]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn = tf.keras.models.Sequential()

# First hidden layer
#nn.add(tf.keras.layers.Dense(unit = hidden_nodes_layer1, input_dim = number_input_features, activation = 'relu'))
nn.add(tf.keras.layers.Dense(units = 1, activation = 'relu', input_dim = 44))

# Second hidden layer
#nn.add(tf.keras.layers.Dense(unit = hidden_nodes_layer2, activation = 'relu'))
nn.add(tf.keras.layers.Dense(units = 1, activation = 'relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units = 1, activation = 'sigmoid'))

# Check the structure of the model
nn.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 1)                 45        
                                                                 
 dense_1 (Dense)             (None, 1)                 2         
                                                                 
 dense_2 (Dense)             (None, 1)                 2         
                                                                 
Total params: 49
Trainable params: 49
Non-trainable params: 0
_________________________________________________________________


In [14]:
# Compile the model
nn.compile(loss = 'binary_crossentropy', optimizer = 'adam', metrics = ['accuracy'])

In [15]:
# Train the model
fit_model = nn.fit(x_train_scaled, y_train, epochs = 100)

Epoch 1/100
804/804 [==============================] - 1s 1ms/step - loss: 0.6625 - accuracy: 0.5892
Epoch 2/100
804/804 [==============================] - 1s 1ms/step - loss: 0.6140 - accuracy: 0.6960
Epoch 3/100
804/804 [==============================] - 1s 993us/step - loss: 0.6000 - accuracy: 0.7034
Epoch 4/100
804/804 [==============================] - 1s 988us/step - loss: 0.5964 - accuracy: 0.7054
Epoch 5/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5942 - accuracy: 0.7086
Epoch 6/100
804/804 [==============================] - 1s 833us/step - loss: 0.5937 - accuracy: 0.7075
Epoch 7/100
804/804 [==============================] - 1s 845us/step - loss: 0.5933 - accuracy: 0.7096
Epoch 8/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5929 - accuracy: 0.7088
Epoch 9/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5927 - accuracy: 0.7095
Epoch 10/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5926 

804/804 [==============================] - 1s 981us/step - loss: 0.5918 - accuracy: 0.7120
Epoch 81/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5919 - accuracy: 0.7118
Epoch 82/100
804/804 [==============================] - 1s 995us/step - loss: 0.5919 - accuracy: 0.7124
Epoch 83/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5918 - accuracy: 0.7119
Epoch 84/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5919 - accuracy: 0.7116
Epoch 85/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5918 - accuracy: 0.7121
Epoch 86/100
804/804 [==============================] - 1s 978us/step - loss: 0.5918 - accuracy: 0.7115
Epoch 87/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5918 - accuracy: 0.7122
Epoch 88/100
804/804 [==============================] - 1s 1ms/step - loss: 0.5917 - accuracy: 0.7121: 0s - l
Epoch 89/100
804/804 [==============================] - 1s 1ms/step - loss: 0.591

In [16]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(x_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.5943 - accuracy: 0.7122 - 304ms/epoch - 1ms/step
Loss: 0.5943447351455688, Accuracy: 0.7121865749359131


In [22]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity_Optimization.h5')

**Overview: 

The goal of the homework was to find the optimal model, to predict the probability of success of charities with the inputs Results. 

Got a model with accuracy of 75%

**Data Preprocessing:

The variable considered the target for this model is the success of the charity (IS_SUCCESSFUL)

**What variable(s) are considered to be the features for your model?

APPLICATION_TYPE
AFFILIATION CLASSIFICATION
USE_CASE
ORGANIZATION STATUS
INCOME_AMT
SPECIAL_CONSIDERATIONS
ASK_AMT

**What variable(s) are neither targets nor features, and should be removed from the input data?

EIN
NAME

**Compiling, Training, and Evaluating the Model:

**How many neurons, layers, and activation functions did you select for your neural network model, and why?

There are three layers the first two layers used relu as the activation function and the outer layer used sigmoid function to try to improve the accuracy of the neural network


**Were you able to achieve the target model performance?

I was not able to get to an accuracy of 75%. The highest I got was around 71%


**What steps did you take to try and increase model performance? 

Found a model that took the inputs but only got a model of around 71%